fix: move to etch and decaffeinate

UziTech opened the pull request:
I recreated the views in etch and decaffeinated the coffeescipt files to javascript so we can use jsx.

I'm not sure what we want to use for eslint config so I just used my standard. I'm ok with anything.

## TODO
- [x] Write Tests
- [x] Agree on eslint config

UziTech committed:
fix: move to etch and decaffeinate

vivi90 commented:
@UziTech Thanks, for your work. I also had in mind to switch over to JavaScript maybe.
@sdinesh86 Is it okay for you, if we switch over from CoffeeScript to JavaScript?

UziTech commented:
We have to move to javascript if we want to use jsx. Atom's version of coffeescript (v1.12.7) doesn't support jsx.

vivi90 commented:
@UziTech Would like to get an "okay" from @sdinesh86, because it's an great change.

vivi90 commented:
@UziTech 
> I'm not sure what we want to use for eslint config so I just used my standard. I'm ok with anything.

I want to get all possible warnings so we can prevent bugs as good as possible.

vivi90 commented on .eslintrc.js:
```diff
@@ -0,0 +1,36 @@
+		}
+	},
+	"rules": {
+		indent: ["error", "tab"],
```
@UziTech Why tab and not 4 spaces?

UziTech replied:
developer accessibility. With tabs the person looking at the code gets to decide how many characters to indent instead of an arbitrary number set by someone else.

I prefer 2 character indentation but I work on so many different projects I don't even notice when it is different. I just press the <kbd>Tab</kbd> key and let Atom decide if it should enter spaces or a tab character.

vivi90 replied:
Yes, Atom decides based on the detected project default.
But i understand it like this is the physical indention.
So i would prefer the default (4 spaces).

UziTech replied:
should I just leave them all as default?

vivi90 commented on .eslintrc.js:
```diff
@@ -0,0 +1,36 @@
+		"jsx-quotes": ["error", "prefer-double"],
+
+		"react/jsx-uses-vars": 2,
+		"react/jsx-indent": ["error", "tab"],
```
@UziTech Same as in line 23.

vivi90 commented on .eslintrc.js:
```diff
@@ -0,0 +1,36 @@
+		quotes: ["error", "double"],
+		"jsx-quotes": ["error", "prefer-double"],
+
+		"react/jsx-uses-vars": 2,
```
@UziTech What means this "2" in the options?
Can't find the documentation for `react/jsx-uses-vars`.

UziTech replied:
`2` = `"error"`

I'll change it.

vivi90 replied:
Ah yes, sorry my fault.

vivi90 commented on lib/atom-music-playlist-view.js:
```diff
@@ -0,0 +1,74 @@
+const etch = require("etch");
+const SelectList = require("atom-select-list");
+
+class PlayListView {
```
@UziTech I think this class needs some refactoring.
For example the name `PlayListView` is misleading, i think.
Maybe something like `SearchView` would be better in my opinion.
And some parts of the class can become optimized.
What do you think about it?

vivi90 replied:
@UziTech Okay, do you want to do these changes in this PR or should we do it later? 
Because it's not critical.

vivi90 commented:
@UziTech We realy need to write more tests. To be honest i am not familiar with Jasmine until now.

UziTech committed:
fix: playlist-view -> search-view

UziTech committed:
test: add search-view tests

UziTech committed:
test: add a view test

UziTech committed:
chore(lint): use default eslint

UziTech commented:
I added a few tests. We can add more later when we fuzz out how to best optimize the classes.

I removed the unnecessary eslint rules so it is only using the recommended settings and the `react/jsx-uses-vars` rule so we can just use the `SelectList` in jsx without eslint thinking it is an unused variable.

sdinesh86 commented:
hello guys sorry for the late response.

@vivi90  I am completely ok with js, honestly I wanted to switch that long time back.

sdinesh86 commented on lib/atom-music-view.js:
```diff
@@ -0,0 +1,344 @@
+		atom.workspace.toggle(this.getURI());
+	}
+
+	changeTicker(e) {
```
Since this is an event handler I would prefer we name it accordingly.
I would suggest something like 

onTickerClicked, onTickerPressed, onTickerPress etc..

vivi90 replied:
@sdinesh86 Yes, your right. Might be a good idea.

sdinesh86 commented on lib/atom-music-view.js:
```diff
@@ -0,0 +1,344 @@
+		}
+	}
+
+	songEnded() {
```
Same as above

sdinesh86 commented on lib/atom-music-view.js:
```diff
@@ -0,0 +1,344 @@
+		}
+	}
+
+	forward15() {
+		this.skip(15);
+	}
+
+	back15() {
+		this.skip(-15);
+	}
```
Do we really need this ?

vivi90 replied:
@sdinesh86 No, not really. But i think it's nice to have, if we tune it a little bit with the ability to change this in the settings.

UziTech replied:
I think this is useful. Occasionally I listen to podcasts and being able to skip back to hear something again is very helpful.

UziTech replied:
having just the command and not a button would be ok with me.

vivi90 replied:
@UziTech Okay, added config setting for it.

sdinesh86 commented on lib/atom-music-view.js:
```diff
@@ -0,0 +1,344 @@
+		}
+	}
+
+	audioPlay() {
```
Since this is an event handler I would prefer we name it accordingly.
I would suggest something like

onAudioPlay etc.

sdinesh86 commented on lib/atom-music-view.js:
```diff
@@ -0,0 +1,344 @@
+		this.startTicker();
+	}
+
+	audioPause() {
```
Same as above

sdinesh86 commented on lib/atom-music.js:
```diff
@@ -0,0 +1,75 @@
+			if (uri === this.atomMusicView.getURI()) { return this.atomMusicView; }
+		}));
+
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:toggle": () => this.atomMusicView.toggle()}));
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:play-pause": () => this.atomMusicView.togglePlayback()}));
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:toggle-shuffle": () => this.atomMusicView.toggleShuffle()}));
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:show-playlist": () => this.atomMusicView.showPlayList()}));
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:forward-15s": () => this.atomMusicView.forward15()}));
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:backward-15s": () => this.atomMusicView.back15()}));
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:next-track": () => this.atomMusicView.nextTrack()}));
+		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:previous-track": () => this.atomMusicView.prevTrack()}));
+
```
Can we refactor this into a new functions

vivi90 replied:
@sdinesh86 Not sure how to do, because it's default to do it this way. But your right, it doesn't look very great.
@UziTech How do you think about this part?

sdinesh86 replied:
```javascript
initializeSubscriptions ( ) {
this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:toggle": () => this.atomMusicView.toggle()}));
		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:play-pause": () => this.atomMusicView.togglePlayback()}));
		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:toggle-shuffle": () => this.atomMusicView.toggleShuffle()}));
		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:show-playlist": () => this.atomMusicView.showPlayList()}));
		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:forward-15s": () => this.atomMusicView.forward15()}));
		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:backward-15s": () => this.atomMusicView.back15()}));
		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:next-track": () => this.atomMusicView.nextTrack()}));
		this.subscriptions.add(atom.commands.add("atom-workspace", {"atom-music:previous-track": () => this.atomMusicView.prevTrack()}));

}

activate(state) {
		this.atomMusicView = new AtomMusicView(state.atomMusicViewState);
		this.subscriptions.add(atom.workspace.addOpener(uri => {
			if (uri === this.atomMusicView.getURI()) { return this.atomMusicView; }
		}));
               this.initializeSubscriptions();
}
```

sdinesh86 replied:
just would make it a bit cleaner and managable

vivi90 replied:
@sdinesh86 Hm, a little bit cleaner but not so much. 
@sdinesh86 @UziTech Would be great, it it's possible to add multiple commands and subscriptions at the same time. But because Atom is not very well documented, i don't know how.

sdinesh86 replied:
I have not been working with atom for quite some time now, so I am not sure about the new api.
But if i remember correctly back then there was a way to add multiple commands at the same time I dont remember it now but if get it I will let you know

UziTech replied:
I updated it to only use a single call to `atom.commands.add`

vivi90 replied:
@UziTech Great! :-)

UziTech committed:
fix: update method names

UziTech committed:
chore: simplify adding commands

UziTech committed:
fix: fix show playlist command

UziTech committed:
test: add command tests

vivi90 committed:
add: skip config

vivi90 committed:
fix: spec test

UziTech commented:
If we are changing commands we should add a notification on the old commands in case anyone uses that command on a toolbar. If we just change commands it will just error out without giving them a clue for what to change it to.

vivi90 commented:
@UziTech Your right, my fault. Where to notice it? In the README file?

vivi90 commented on lib/atom-music-view.js:
```diff
@@ -238,7 +238,7 @@ class AtomMusicView {
 	}
 
-	filesBrowsed(e) {
+	onFilesChanged(e) {
```
@UziTech `onFilesBrowsed` might be more clear.

UziTech replied:
This is called when the "files" input changed. `on={{change: this.onFilesChanged}} `

I don't think "browsed" would make it any more clear.

vivi90 replied:
Yes, `files` changes. 
But the described process is browsing for files to add.
`..Changed` might sound a little bit like observing files for changes, i think.

UziTech replied:
Maybe `onPlaylistChange`? I feel like the names of these functions aren't really important enough to need to be changed. This function, for example, is only used in one place. It is not difficult for anyone to know what it does no matter what name is used.

vivi90 replied:
@sdinesh86 What do you think?

vivi90 replied:
@UziTech I will change it to `onFilesAdded`.

UziTech commented:
I would add the commands back and add a notification with [`atom.notifications.addWarning()`](https://flight-manual.atom.io/api/v1.47.0/NotificationManager/#instance-addWarning) saying something about deprecating that command and changing to the new command. and maybe in a month or two remove the command.

UziTech commented:
Before we change this too much can we merge this PR and add other changes as other PRs?

vivi90 commented:
@UziTech How about `addInfo`?
Hm, let's also add it to the README file or documentation later.

sdinesh86 commented:
So should I go ahead and merge the PR ?

sdinesh86 commented:
Also guys you can take the liberty to merge the PR's as needed I really doubt I will get enough time in next few week's to start working on this.

vivi90 commented:
@UziTech @sdinesh86 Okay so i will change some little things and then merge it.
Some details can discussed and changed later, if needed.

vivi90 commented:
@UziTech @sdinesh86 I will increase the major version, because it might be API-breaking because of so much changes, i think.

vivi90 reviewed (approved):
Okay, will become merged.

vivi90 commented:
Fixes #25
